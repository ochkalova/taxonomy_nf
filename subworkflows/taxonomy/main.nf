include { DIAMOND_BLASTP } from '../../modules/nf-core/diamond/blastp/main'
include { TAXONKIT_LCA   } from '../../modules/nf-core/taxonkit/lca/main'

process REFORMAT_DIAMOND_OUTPUT {

    input:
    tuple val(meta), path(diamond_file)

    output:
    tuple val(meta), path("contig_to_taxid.tsv"), emit: converted_file

    script:
    
    """
    #!/bin/bash

    # Input and output files
    input_file="${diamond_file}"
    output_file="contig_to_taxid.tsv"

    # Initialize variables
    prev_test_id=""
    taxids=""

    # Read the input file line by line
    while read line; do
    # Extract the test ID (the part before the last underscore in the first column)
    test_id=\$(echo "\$line" | cut -f1 | sed 's/_[^_]*\$//')

    # Extract the TaxID using awk and grep for TaxID field
    taxid=\$(echo "\$line" | grep -o "TaxID=[0-9]*" | cut -d'=' -f2)

    # If we're still on the same test ID, append the TaxID
    if [[ \$test_id" == "\$prev_test_id" || "\$prev_test_id" == "" ]]; then
        if [[ "\$taxids" != "" ]]; then
        taxids="\$taxids,\$taxid"
        else
        taxids="\$taxid"
        fi
    else
        # If test ID changed, print the previous result and reset
        echo "\$prev_test_id   \$taxids" >> "\$output_file"
        taxids="\$taxid"
    fi

    prev_test_id="\$test_id"
    done < "\$input_file"

    # Print the last test ID and its TaxIDs
    echo "\$prev_test_id   \$taxids" >> "\$output_file"
    """
}

workflow TAXONOMY {
    take:
    proteins                    // [ val(meta), path(assembly_fasta) ]

    main:
    
    db = [[id: ""], params.taxonomy_db]
    DIAMOND_BLASTP(proteins, db, "txt", "qseqid salltitles")

    REFORMAT_DIAMOND_OUTPUT(DIAMOND_BLASTP.out.txt)

    TAXONKIT_LCA(REFORMAT_DIAMOND_OUTPUT.out.converted_file, 1, ",")

    emit:
    output = TAXONKIT_LCA.out.converted_file
}
