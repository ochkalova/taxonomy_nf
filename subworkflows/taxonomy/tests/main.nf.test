nextflow_workflow {

    name "Test subworkflows/taxonomy/tests/main.nf.test"
    script "../main.nf"
    workflow "TAXONOMY"

    test("Run the pipeline on the test dataset") {
        when {
            params {
                output_path    = "${outputDir}"
            }
            workflow {
                """
                input[0] = [ [id:'test'], file("${moduleDir}/tests/data/contigs.fasta", checkIfExists: true) ]
                input[1] = [ [id:'test'], file("${moduleDir}/tests/data/proteins.fasta", checkIfExists: true) ]
                input[2] = [ [id:'test'], file("${moduleDir}/tests/data/db", checkIfExists: true) ]
                input[3] = [ [id:'test'], file("${moduleDir}/tests/data/tax", checkIfExists: true) ]
                """
            }
        }
        then {
            assertAll(
            { assert workflow.success },
            { assert snapshot(workflow.out.diamond_output).match("diamond_output") },
            { assert snapshot(workflow.out.cat_output).match("cat_output") },
            )
        }
    }
}
