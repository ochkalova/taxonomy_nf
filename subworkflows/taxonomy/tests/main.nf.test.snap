{
    "cat_output": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.contig2classification.txt:md5,3c3c79045bf6ae8b1292ae9afa2ec4af"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-01-29T11:41:47.358582"
    },
    "diamond_output": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.txt:md5,9e2f9c188b183c18dd9572395a48a066"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.0"
        },
        "timestamp": "2025-01-29T11:41:47.351764"
    }
}